docs/dev-guide/contrib-guides-cpp.md

```diff
@@ -0,0 +1,158 @@
+# C++
+
+Follow the guidelines below when writing and updating C++ files.
+
+## Automated linting
+
+As mentioned in the overview, where possible, we have automated linting processes, so you need not
+remember all the guidelines perfectly. For C++, we currently use the following tools:
+
+* `clang-format` for formatting ([base config][clang-format-config])
+* `clang-tidy` for static analysis ([base config][clang-tidy-config])
+
+When the linter disagrees with a guideline, err on the side of following the linter, since:
+
+* we don't want automated runs of the linters to fail, leading to them being ignored due to the
+  noise.
+* the linter may have a good reason for disagreeing with the guideline.
+* the linter's configuration may be more up-to-date than the guideline.
+
+:::{tip}
+To learn about how to apply the linters to a new C++ project, [see here][adding-cpp-linting].
+:::
+
+When you encounter such a disagreement, if it hasn't been noted below, please open an issue to track
+it.
+
+## Guidelines
+
+We adhere to [Google's C++ style guide][google-cpp-style-guide] (as of
+[8f97e24][google-styleguide-8f97e24]) with the following exceptions (organized according to the
+sections in Google's style guide).
+
+:::{note}
+This section is a work in progress and does not yet include all exceptions after the
+[Classes][google-cpp-style-guide-classes] section of Google's style guide.
+:::
+
+### Header files
+
+#### Self-contained Headers
+
+* Header files should end in `.hpp`.
+* Don't use non-header files meant for inclusion, since they can confuse static analysis tools.
```

Comment:
I feel there may be a chance someone unfamiliar with c++ may not understand "non-header files meant for inclusion" and the relationship to "self-contained" headers. Maybe we could add a bracket example like::
> ... non-header files meant for inclusion (e.g. `.impl` files containing definitions directly included at the end of the `.hpp` file), since they ...